cleverhans/picklable_model.py

```diff
@@ -0,0 +1,372 @@
+            if not hasattr(layers[-1], 'name'):
+                layers[-1].name = 'probs'
+            if not hasattr(layers[-2], 'name'):
+                layers[-2].name = 'logits'
+        else:
+            if not hasattr(layers[-1], 'name'):
+                layers[-1].name = 'logits'
+        for i, layer in enumerate(self.layers):
+            if layer.parent is None:
+                if i == 0:
+                    layer.parent = "input"
+                else:
+                    layer.parent = layers[i - 1].name
+            if hasattr(layer, 'name'):
+                name = layer.name
+            else:
+                name = layer.__class__.__name__ + str(i)
+                layer.name = name
+            self.layer_names.append(name)
+
+            layer.set_input_shape(input_shape)
+            input_shape = layer.get_output_shape()
+
+    def get_params(self):
+        out = []
+        for layer in self.layers:
+            out = ordered_union(out, layer.get_params())
+        return out
+
+    def fprop(self, x=None, set_ref=False, given=None):
+
+        # Note: this currently isn't great.
+        # A layer can have any parent it wants, but the parent
+        # must come earlier in the list.
+        # There's no way to have > 1 parent.
+        # This means we can support branched structures that split,
+        # e.g. for multiple output heads, but not structures
+        # that converge.
+        # We can feed a value in the middle using "given" but
+        # only layers after the given one are run using the current
+        # implementation, so the feed must happen before any branch
+        # point.
+
+        if x is None:
+            if given is None:
+                raise ValueError("One of `x` or `given` must be specified")
+        else:
+            assert given is None
+            given = ('input', x)
+        name, value = given
+        out = {name: value}
+        x = value
+
+        if name == 'input':
+            layers = self.layers
+        else:
+            for i, layer in enumerate(self.layers[:-1]):
+                if layer.name == name:
+                    layers = self.layers[i+1:]
+                    break
+
+        for layer in layers:
+            x = out[layer.parent]
+            if set_ref:
+                layer.ref = x
+            x = layer.fprop(x)
+            assert x is not None
+            out[layer.name] = x
+        return out
+
+
+class Layer(PicklableModel):
+    def __init__(self, name=None, parent=None):
+        if name is not None:
+            self.name = name
+        self.parent = None
+
+    def get_output_shape(self):
+        return self.output_shape
+
+
+class Linear(Layer):
+
+    def __init__(self, num_hid, init_scale=1., init_b=0., **kwargs):
+        super(Linear, self).__init__(**kwargs)
+        self.num_hid = num_hid
+        self.init_scale = init_scale
+        self.init_b = init_b
+
+    def set_input_shape(self, input_shape):
```

Comment:
It's true that I do most of the initialization inside this method for most of the layers I've implemented, but I don't think we need to make that a requirement of the interface. I suspect that when we start wrapping models written by other people we'll find that they do almost all the initialization inside the `fprop` method and that the model just exists in an invalid state until `fprop` is called. That's allowed by the current interface (`get_params` will raise an exception if you call it before `fprop` on models that use that strategy, but that's allowed).

Reply:
I see, I guess this would be kind of like the few lines that you added here: https://github.com/tensorflow/cleverhans/pull/507/files#diff-597d409a81b3230e149314553adbd67bR24